enable fp16 autocast for dynamo benchmark

zhuhaozhe opened the pull request:
`--amp` to enable amp path for` CUDA` (default amp_dtype will be float16) and `CPU` (default amp_dtype will be bfloat16).

If users set `--amp_dtype`, the amp_dtype from users will have the highest priority.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114088



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zhuhaozhe committed:
enable fp16 autocast for dynamo benchmark

[ghstack-poisoned]

jgong5 commented on benchmarks/dynamo/common.py:
```diff
@@ -1977,6 +1978,10 @@ def setup_amp(self):
             self.autocast = torch.cuda.amp.autocast
         elif (self.args.bfloat16 or self.args.amp) and self.args.devices == ["cpu"]:
             self.autocast = torch.cpu.amp.autocast
+            amp_dtype = (
+                torch.float16 if self.args.amp_dtype == "float16" else torch.bfloat16
+            )
+            self.autocast_arg["dtype"] = amp_dtype
```
Do we still need to check `self.args.bfloat16`? Can we align the logic between CPU and CUDA considering both support the same set of amp types? Also please note that CPU and CUDA have different default amp dtype when the dtype is not provided to the autocast API.

zhuhaozhe replied:
Hi, @jgong5. What I considered is not to impact the user's existing scripts. For example, the user may set `--dtype bfloat16` without setting `--amp`. The original behavior will still run amp).

jgong5 replied:
What's the behavior on CUDA? Is this original behavior aligned with CUDA?

zhuhaozhe replied:
Hi, @jgong5. For CUDA, the behavior before/after this PR:
Set `--amp` to enable `autocast with fp16` (since fp16 is cuda default)`.

The behavior before this PR is not aligned between CPU and CUDA (CPU can enable BF16 autocast with `--amp` **or**
`--dtype bfloat16`.  So if we wish to align CPU/CUDA behavior with autocast in this PR, we may need to change the behavior for CPU or CUDA.

zhuhaozhe replied:
Hi, @jgong5. Got it, now they are aligned.
--amp will enable amp path for CUDA (default amp_dtype will be float16) and CPU (default amp_dtype will be bfloat16).
If user further set --amp_dtype, the amp_dtype will use the dtype set by users.

zhuhaozhe committed:
Update on "enable fp16 autocast for dynamo benchmark"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on benchmarks/dynamo/common.py:
```diff
@@ -2965,7 +2974,11 @@ def get_example_inputs(self):
     )
-
+    parser.add_argument(
+        "--amp_dtype",
```
```suggestion
        "--amp-dtype",
```

zhuhaozhe replied:
Thanks for the advice, @jansel.

zhuhaozhe committed:
Update benchmarks/dynamo/common.py

Co-authored-by: Jason Ansel <jansel@meta.com>

zhuhaozhe committed:
Update common.py

zhuhaozhe committed:
Update on "enable fp16 autocast for dynamo benchmark"

`--amp` to enable amp path for` CUDA` (default amp_dtype will be float16) and `CPU` (default amp_dtype will be bfloat16).

If users set `--amp_dtype`, the amp_dtype from users will have the highest priority.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zhuhaozhe commented on benchmarks/dynamo/common.py:
```diff
@@ -1759,38 +1759,47 @@ def __init__(self):
-        elif (self.args.bfloat16 or self.args.amp) and self.args.devices == ["cpu"]:
-            self.autocast = torch.cpu.amp.autocast
+        if self.args.amp:
+            if self.args.devices == ["cuda"]:
```
Hi, @jansel.
I extend `setup_amp` to have an arg for `current_device` and invoke it after getting `current_device` before `run_model`.

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot merge
